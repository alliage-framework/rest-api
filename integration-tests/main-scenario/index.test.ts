import fs from "fs";

import { Sandbox } from "@alliage/sandbox";
import { WebserverSandbox } from "@alliage/webserver-sandbox";

describe("Main scenario", () => {
  const sandbox = new Sandbox({
    scenarioPath: __dirname,
  });

  const METADATA_PATH = `${sandbox.getPath()}/.alliage-rest-api-metadata.json`;

  const webserverSandbox = new WebserverSandbox(sandbox);

  beforeAll(async () => {
    try {
      await sandbox.init();
    } catch (e) {
      console.error(e);
    }
  });

  afterAll(async () => {
    await sandbox.clear();
  });

  describe("Schema generation", () => {
    beforeEach(() => {
      if (fs.existsSync(METADATA_PATH)) {
        fs.unlinkSync(METADATA_PATH);
      }
    });

    describe("With process", () => {
      it("should generate a schema from the controllers types", async () => {
        const p = await sandbox.run(["rest:generate-schema"]);
        await p.waitCompletion();

        const schema = fs.readFileSync(METADATA_PATH).toString();

        expect(JSON.parse(schema)).toMatchSnapshot();
      }, 8000);
    });

    describe("With builder task", () => {
      it("should generate a schema from the controllers types", async () => {
        const p = await sandbox.build([]);
        await p.waitCompletion();

        const schema = fs.readFileSync(METADATA_PATH).toString();

        expect(JSON.parse(schema)).toMatchSnapshot();
      }, 8000);
    });
  });

  describe("Rest API", () => {
    beforeAll(async () => {
      try {
        await webserverSandbox.start();
      } catch (e) {
        console.error(e);
        process.exit(-1);
      }
    }, 8000);

    afterAll(async () => {
      await webserverSandbox.stop();
    });

    it("should provide a OpenAPI spec", async () => {
      const res = await webserverSandbox.getClient().get("/api/specs");

      expect(res.status).toEqual(200);
      expect(res.data).toMatchSnapshot();
    });

    it("should say hello in french", async () => {
      const res = await webserverSandbox
        .getClient()
        .post("/api/hello/Jean?language=fr", { age: 42 });

      expect(res.status).toEqual(200);
      expect(res.data).toEqual({ message: "Bonjour Jean, tu as 42 ans" });
    });

    it("should say hello in english", async () => {
      const res = await webserverSandbox
        .getClient()
        .post("/api/hello/John?language=en", { age: 26 });

      expect(res.status).toEqual(200);
      expect(res.data).toEqual({ message: "Hello John, you are 26" });
    });

    it("should return an error when the body is invalid", async () => {
      const res = await webserverSandbox
        .getClient()
        .post(
          "/api/hello/John?language=fr",
          { age: "twenty-six" },
          { validateStatus: () => true }
        );

      expect(res.status).toEqual(400);
      expect(res.data).toEqual([
        {
          errors: [
            {
              instancePath: "/age",
              keyword: "type",
              message: "must be number",
              params: { type: "number" },
              schemaPath: "#/properties/age/type",
            },
          ],
          source: "body",
        },
      ]);
    });

    it("should return an error when the params are invalid", async () => {
      const res = await webserverSandbox
        .getClient()
        .post(
          "/api/hello/42?language=fr",
          { age: 26 },
          { validateStatus: () => true }
        );

      expect(res.status).toEqual(400);
      expect(res.data).toEqual([
        {
          errors: [
            {
              instancePath: "/name",
              keyword: "pattern",
              message: 'must match pattern "[a-zA-Z]+"',
              params: { pattern: "[a-zA-Z]+" },
              schemaPath: "#/properties/name/pattern",
            },
          ],
          source: "params",
        },
      ]);
    });

    it("should return an error when the query is invalid", async () => {
      const res = await webserverSandbox
        .getClient()
        .post(
          "/api/hello/John?language=it",
          { age: 26 },
          { validateStatus: () => true }
        );

      expect(res.status).toEqual(400);
      expect(res.data).toEqual([
        {
          errors: [
            {
              instancePath: "/language",
              keyword: "enum",
              message: "must be equal to one of the allowed values",
              params: { allowedValues: ["fr", "en"] },
              schemaPath: "#/properties/language/enum",
            },
          ],
          source: "query",
        },
      ]);
    });

    it("should handle preflight requests", async () => {
      const res = await webserverSandbox
        .getClient()
        .options("/api/hello/John", {
          headers: {
            "Access-Control-Request-Method": "",
            "Access-Control-Request-Headers": "",
            Origin: "http://acme.com",
          },
        });

      expect(res.status).toBe(204);
      expect(res.headers).toEqual(
        expect.objectContaining({
          "access-control-allow-origin": "http://acme.com",
          "access-control-allow-headers":
            "X-Custom-Header-1, X-Custom-Header-2",
          "access-control-allow-methods": "GET, POST, PUT, DELETE",
          "access-control-max-age": "4800",
        })
      );
    });
  });
});
