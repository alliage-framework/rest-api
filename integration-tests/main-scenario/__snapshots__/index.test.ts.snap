// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main scenario Rest API should provide a OpenAPI spec 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "support@example.com",
      "name": "API Support",
      "url": "http://www.example.com/support",
    },
    "description": "Describe what your API does here...",
    "license": Object {
      "name": "Apache 2.0",
      "url": "https://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "title": "Alliage REST API",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/api/hello/:name": Object {
      "post": Object {
        "parameters": Object {
          "language": Object {
            "in": "query",
            "required": false,
            "schema": Object {
              "enum": Array [
                "fr",
                "en",
              ],
              "type": "string",
            },
          },
          "name": Object {
            "in": "path",
            "required": true,
            "schema": Object {
              "pattern": "[a-zA-Z]+",
              "type": "string",
            },
          },
        },
        "requestBody": Object {
          "additionalProperties": false,
          "properties": Object {
            "age": Object {
              "type": "number",
            },
          },
          "required": Array [
            "age",
          ],
          "type": "object",
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "additionalProperties": false,
                  "properties": Object {
                    "message": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "message",
                  ],
                  "type": "object",
                },
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`Main scenario Schema generation With builder task should generate a schema from the controllers types 1`] = `
Object {
  "post": Array [
    Object {
      "actionMetadata": Object {
        "bodyType": Object {
          "additionalProperties": false,
          "properties": Object {
            "age": Object {
              "type": "number",
            },
          },
          "required": Array [
            "age",
          ],
          "type": "object",
        },
        "controllerName": "MainController",
        "defaultStatusCode": 200,
        "errors": Array [],
        "name": "sayHello",
        "paramsType": Object {
          "additionalProperties": false,
          "properties": Object {
            "name": Object {
              "pattern": "[a-zA-Z]+",
              "type": "string",
            },
          },
          "required": Array [
            "name",
          ],
          "type": "object",
        },
        "queryType": Object {
          "additionalProperties": false,
          "properties": Object {
            "language": Object {
              "enum": Array [
                "fr",
                "en",
              ],
              "type": "string",
            },
          },
          "type": "object",
        },
        "returnType": Object {
          "additionalProperties": false,
          "properties": Object {
            "message": Object {
              "type": "string",
            },
          },
          "required": Array [
            "message",
          ],
          "type": "object",
        },
        "validateInput": true,
        "validateOutput": true,
      },
      "path": "/api/hello/:name",
      "pattern": "/^\\\\/api\\\\/hello(?:\\\\/([^\\\\/#\\\\?]+?))[\\\\/#\\\\?]?$/i",
    },
  ],
}
`;

exports[`Main scenario Schema generation With process should generate a schema from the controllers types 1`] = `
Object {
  "post": Array [
    Object {
      "actionMetadata": Object {
        "bodyType": Object {
          "additionalProperties": false,
          "properties": Object {
            "age": Object {
              "type": "number",
            },
          },
          "required": Array [
            "age",
          ],
          "type": "object",
        },
        "controllerName": "MainController",
        "defaultStatusCode": 200,
        "errors": Array [],
        "name": "sayHello",
        "paramsType": Object {
          "additionalProperties": false,
          "properties": Object {
            "name": Object {
              "pattern": "[a-zA-Z]+",
              "type": "string",
            },
          },
          "required": Array [
            "name",
          ],
          "type": "object",
        },
        "queryType": Object {
          "additionalProperties": false,
          "properties": Object {
            "language": Object {
              "enum": Array [
                "fr",
                "en",
              ],
              "type": "string",
            },
          },
          "type": "object",
        },
        "returnType": Object {
          "additionalProperties": false,
          "properties": Object {
            "message": Object {
              "type": "string",
            },
          },
          "required": Array [
            "message",
          ],
          "type": "object",
        },
        "validateInput": true,
        "validateOutput": true,
      },
      "path": "/api/hello/:name",
      "pattern": "/^\\\\/api\\\\/hello(?:\\\\/([^\\\\/#\\\\?]+?))[\\\\/#\\\\?]?$/i",
    },
  ],
}
`;
